lib/remote_node.go

```diff
@@ -0,0 +1,599 @@
+package lib
+
+import (
+	"encoding/binary"
+	"fmt"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/deso-protocol/core/bls"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/sha3"
+	"net"
+	"sync"
+	"time"
+)
+
+type RemoteNodeStatus int
+
+const (
+	RemoteNodeStatus_NotConnected RemoteNodeStatus = 0
+	RemoteNodeStatus_Connected    RemoteNodeStatus = 1
+	RemoteNodeStatus_Validated    RemoteNodeStatus = 2
+	RemoteNodeStatus_Attempted    RemoteNodeStatus = 3
+	RemoteNodeStatus_Terminated   RemoteNodeStatus = 4
+)
+
+type RemoteNodeId uint64
+
+func NewRemoteNodeId(id uint64) RemoteNodeId {
+	return RemoteNodeId(id)
+}
+
+func (id RemoteNodeId) ToUint64() uint64 {
+	return uint64(id)
+}
+
+// RemoteNode is a consensus-aware wrapper around the network Peer object. It is used to manage the lifecycle of a peer
+// and to store consensus-related metadata about the peer. The RemoteNode can wrap around either an inbound or outbound
+// peer connection. For outbound peers, the RemoteNode is created prior to the connection being established. In this case,
+// the RemoteNode will be first used to initiate an OutboundConnectionAttempt, and then store the resulting connected peer.
+// For inbound peers, the RemoteNode is created after the connection is established in ConnectionManager.
+//
+// Once the RemoteNode's peer is set, the RemoteNode is used to manage the handshake with the peer. The handshake involves
+// rounds of Version and Verack messages being sent between our node and the peer. The handshake is complete when both
+// nodes have sent and received a Version and Verack message. Once the handshake is successful, the RemoteNode will
+// emit a MsgDeSoPeerHandshakeComplete control message via the Server.
+//
+// In steady state, i.e. after the handshake is complete, the RemoteNode can be used to send a message to the peer,
+// retrieve the peer's handshake metadata, and close the connection with the peer. The RemoteNode has a single-use
+// lifecycle. Once the RemoteNode is terminated, it will be disposed of, and a new RemoteNode must be created if we
+// wish to reconnect to the peer in the future.
+type RemoteNode struct {
+	mtx sync.Mutex
+
+	peer *Peer
+	// The id is the unique identifier of this RemoteNode. For outbound connections, the id will be the same as the
+	// attemptId of the OutboundConnectionAttempt, and the subsequent id of the outbound peer. For inbound connections,
+	// the id will be the same as the inbound peer's id.
+	id               RemoteNodeId
+	connectionStatus RemoteNodeStatus
+
+	params *DeSoParams
+	srv    *Server
+	bc     *Blockchain
```

Comment:
Try removing the blockchain field from RemoteNode. Seems like it's only used to extract the chain's current height when sending the version message. The height can be passed as a param into InitiateHandshake and HandleVersionMessage by the caller instead. This will make sure the RemoteNode has less knowledge of the chain
```suggestion
```

Reply:
agreed